scp/scpProcessCreator.cpp

```diff
@@ -1,105 +1,120 @@
 #include "scpKeynodes.hpp"
 #include "scpUtils.hpp"
-#include "scpProcessCreator.hpp"
-#include "sc-memory/sc_memory.hpp"
-#include <iostream>
-
-namespace scp {
-ScAddr ASCPProcessCreator::msAgentKeynode;
 
-SC_AGENT_IMPLEMENTATION(ASCPProcessCreator)
+namespace scp
 {
-    if (!edgeAddr.IsValid())
-        return SC_RESULT_ERROR;
-
-    ScAddr action =m_memoryCtx.GetEdgeTarget(edgeAddr);
-
-    if (!m_memoryCtx.HelperCheckEdge(Keynodes::action_scp_interpretation_request, action, ScType::EdgeAccessConstPosPerm))
-        return SC_RESULT_ERROR_INVALID_PARAMS;
-
-    ScAddr program, params;
-    ScIterator5Ptr iter_param =m_memoryCtx.Iterator5(action, ScType::EdgeAccessConstPosPerm, ScType::Node, ScType::EdgeAccessConstPosPerm, Keynodes::rrel_1);
-    if (iter_param->Next())
-        program = iter_param->Get(2);
-    else
-        return SC_RESULT_ERROR_INVALID_PARAMS;
-
-    iter_param =m_memoryCtx.Iterator5(action, ScType::EdgeAccessConstPosPerm, ScType::Node, ScType::EdgeAccessConstPosPerm, Keynodes::rrel_2);
-    if (iter_param->Next())
-        params = iter_param->Get(2);
-    else
-        return SC_RESULT_ERROR_INVALID_PARAMS;
+ScResult ASCPProcessCreator::DoProgram(
+    ScEventAfterGenerateOutgoingArc<ScType::EdgeAccessConstPosPerm> const & event,
+    ScAction & action)
+{
+  auto const & [program, params] = action.GetArguments<2>();
+  if (!program.IsValid() || !params.IsValid())
+    return action.FinishUnsuccessfully();
 
-    ScTemplate program_templ;
-   m_memoryCtx.HelperBuildTemplate(program_templ, program);
-    ScTemplateParams gen_params;
+  ScTemplate program_templ;
+  m_context.HelperBuildTemplate(program_templ, program);
+  ScTemplateParams gen_params;
 
-    ScAddr process_node;
-    ScIterator5Ptr iter_temp =m_memoryCtx.Iterator5(program, ScType::EdgeAccessConstPosPerm, ScType::NodeVar, ScType::EdgeAccessConstPosPerm, Keynodes::rrel_key_sc_element);
-    if (iter_temp->IsValid() && iter_temp->Next())
-        process_node = iter_temp->Get(2);
-    else
-        return SC_RESULT_ERROR_INVALID_PARAMS;
+  ScAddr process_node;
+  auto const & programKeyElementIterator = m_context.Iterator5(
+      program,
+      ScType::EdgeAccessConstPosPerm,
+      ScType::NodeVar,
+      ScType::EdgeAccessConstPosPerm,
+      Keynodes::rrel_key_sc_element);
+  if (programKeyElementIterator->IsValid() && programKeyElementIterator->Next())
+    process_node = programKeyElementIterator->Get(2);
+  else
+    return action.FinishUnsuccessfully();
 
-    iter_temp =m_memoryCtx.Iterator5(process_node, ScType::EdgeAccessVarPosPerm, ScType::NodeVar, ScType::EdgeAccessConstPosPerm, program);
-    if (!iter_temp->IsValid())
-        return SC_RESULT_ERROR_INVALID_PARAMS;
+  auto const & processParametersIterator = m_context.Iterator5(
+      process_node, ScType::EdgeAccessVarPosPerm, ScType::Unknown, ScType::EdgeAccessConstPosPerm, program);
+  if (!processParametersIterator->IsValid())
+  {
+    return action.FinishUnsuccessfully();
```

Comment:
Log error please

Reply:
I removed  that, I don't believe that iterator can be invalid